### DISCUSSION TITLE
If I got it right, is mir similar to LLVM?

### DISCUSSION THREAD

**Discussion by rempas** [vnmakarov/mir#207] [Q&A]
I've read the README and also the blog posts and from my understanding, MIR is an alternative to LLVM right? Then if I'm right, c2m is a C fronted that uses the MIR as a backend.  So just like the LLVM does with Clang? If that's right, then can I create my own fronted and implement my language (or modify the current C one)?

Maybe I'm not right in everything so a little bit of clarification would be nice

**Comment by vnmakarov** [comment]
Yes, you are mostly right.

The difference is that LLVM is a huge compiler framework implementing huge number of optimizations (hundreds of them ) and supporting more targets and MIR-generator implements small number (less than 10) of optimizations but most valuable ones.

Another important difference MIR generates machine code in memory and LLVM generates assembler and/or binary object format (elf/macho).  Although in some way MIR could be considered as a custom *machine independent* object format.  In the future MIR-generator might generate assembler code too.

`c2m` can be considered as simplified analog of clang.  Clang supports C++ and a lot of C extensions, c2m supports only C11 and few C extensions.

You can write own (language) front-end generating MIR and through MIR-generator machine code for it.

**Reply by rempas** [reply]
That's amazing seriously!!! II wanted to make a language front end for LLVM but I was worried about the compilation times and the fact that it is written in C++ so later when I want to bootstrap, I would have to support C++ interpolation (which is... aaaaah). So I think MIR will just be the right tool to do the work.  I would like to make some more questions if you don't mind.

1. You said that MIR generates machine code in memory. First what is that? Can I get any source? Also will it effect runtime performance?
2. Do you consider making more optimizations for when working with numbers in the future? For example when using a program to calculate [fib](https://www.tutorialspoint.com/data_structures_algorithms/fibonacci_recursive_program_in_c.htm) numbers, I find c2m to be 130% slower than optimized (-Os) gcc and 50% slower than Golang (which is surprisingly fast here). So I'm just wondering and your opinion is on this one

**Reply by vnmakarov** [reply]
>     1. You said that MIR generates machine code in memory. First what is that? Can I get any source? Also will it effect runtime performance?
> 

Direct machine code generation from MIR in memory means there are no any temporary object files (.o) or dynamic libraries (.so, .dll) and 
no system linker or system dynamic loader is used.  Of course, code generation is faster in this case.  For example, libgccjit generates assembler files, then .so files, then loads the .so files to call the generated machine code.
 
>     2. Do you consider making more optimizations for when working with numbers in the future? For example when using a program to calculate [fib](https://www.tutorialspoint.com/data_structures_algorithms/fibonacci_recursive_program_in_c.htm) numbers, I find c2m to be 130% slower than optimized (-Os) gcc and 50% slower than Golang (which is surprisingly fast here). So I'm just wondering and your opinion is on this one

Fibonacci is the worst case for MIR-generated code performance.  All calls in MIR are done through small thunks (it means additional 2-4 machine insns for each call).  Such approach is used for MIR to implement hot patching.  You can generate several versions of one function, for example functions with speculative assumptions and if the assumptions were wrong to generate and use deoptimized version.  MIR was originally designed for JITs and speculation/deoptimization is important for JIT of dynamic programming languages.

**Reply by rempas** [reply]
> Of course, code generation is faster in this case

WHAT????? It is faster then than the Binary format GCC/Clang outputs??? It is also smaller in size (the typical "Hello world program" resulted in  3.5 times smaller binary!)! Is this black magic? Also, does not creating object files, libraries and linking means that we will see a significant compilation boost when using multiple files compared to when using one where a saw a speed up of almost 4 times?

> All calls in MIR are done through small thunks (it means additional 2-4 machine insns for each call).

Does that also include jump in labels or only function calls? Also I did indeed tried fib with the [following](https://qawithexperts.com/article/c-cpp/fibonacci-series-program-in-c-with-and-without-recursion/300) and indeed it has performance in par with "gcc -Os" (some times similar, some times worse, some times a little bit better). That's insane!!!!

Thanks a lot for your hard work and for the time you gave on this one! I'm really impressed and I think MIR is what I was looking for! I only have one last thing to ask and that is, which one is the directory that includes that C2m front-end and how is it structured? Is everything about the front-end there so can I just modify the code and create my own front-end as simple as it sounds?

**Reply by vnmakarov** [reply]
> > Of course, code generation is faster in this case
> 
> WHAT????? It is faster then than the Binary format GCC/Clang outputs??? It is also smaller in size (the typical "Hello world program" resulted in 3.5 times smaller binary!)! Is this black magic? Also, does not creating object files, libraries and linking means that we will see a significant compilation boost when using multiple files compared to when using one where a saw a speed up of almost 4 times?
> 

I don't know where did you get these numbers.  I did some benchmarking and posted it on https://developers.redhat.com/sites/default/files/blog/2020/11/commet-lake-compilation-speed.png.  Still tcc is much faster than c2m in overall and in code generation part (although tcc generated code performance is worse than c2m one).

What I meant code generation is faster under the same other conditions because you don't need intermediate formats (object files, dynamic libraries) and their writing/reading.

> > All calls in MIR are done through small thunks (it means additional 2-4 machine insns for each call).
> 
> Does that also include jump in labels or only function calls? Also I did indeed tried fib with the [following](https://qawithexperts.com/article/c-cpp/fibonacci-series-program-in-c-with-and-without-recursion/300) and indeed it has performance in par with "gcc -Os" (some times similar, some times worse, some times a little bit better). That's insane!!!!
>
MIR call on x86-64 looks like
```
calling function code:               thunk:
call -----------------------> movabs reg,<function address>; jmp *reg --------> called function start
```
Function address is changed everytime a new version of function is loaded. 

 
> Thanks a lot for your hard work and for the time you gave on this one! I'm really impressed and I think MIR is what I was looking for! I only have one last thing to ask and that is, which one is the directory that includes that C2m front-end and how is it structured? Is everything about the front-end there so can I just modify the code and create my own front-end as simple as it sounds?

Thank you for your interest in the project.  MIR-project has some advantages and disadvantages.  I'd do some things differently  if I started the project now.  For example, I'd like to improve speed of c2m compilation and machine code generation.

c2m sources are in directory c2mir.  Some information about source structure can be found in https://github.com/vnmakarov/mir/blob/master/HOW-TO-PORT-MIR.md

**Reply by rempas** [reply]
> I don't know where did you get these numbers

I've seen the benchmarks, I tried some small programs and that were the results for my machine but I suppose the bigger the program, the more the compilation speed gain will be.

> I'd do some things differently if I started the project now

That's interesting and if you even feel like talking about that, It would be awesome to write a new discussion. I think other people will be interested to listen too! In the end why not making the project from the beginning  if it's worth it? 😂 

 > I'd like to improve speed of c2m compilation and machine code generation.

Hah?  The code generation that c2m does? Isn't c2m a front-end to MIR? Isn't MIR the one does code generation (hence MIR-generator)? Maybe I got the meaning wrong here...

> c2m sources are in directory c2mir

Thanks a lot! I would try to "play" with it because seriously I can only play at this point that I am (and this at best) and If I make anything worth showing, I would send it idk...

**Reply by vnmakarov** [reply]
> Hah? The code generation that c2m does? Isn't c2m a front-end to MIR? Isn't MIR the one does code generation (hence MIR-generator)? Maybe I got the meaning wrong here...

C to MIR compiler is in source file `c2mir/c2mir.c`.  Machine code generation is provided by MIR-generator (`mir-gen.c`) which is called from `c2mir/c2mir-driver.c`.  C Compiler to machine code `c2m` is built from `c2mir.c`, `c2mir-driver.c`, `mir.c`, and `mir-gen.c`.  

`c2m` can produce MIR-code (binary or textual one) or machine code in memory and `c2m` can execute the machine code.  `c2m` can take C code or/and MIR binary or textual form.  So `c2m` can do a lot.

**Reply by data-man** [reply]
@vnmakarov 
Has libmir an analogous to `llvm.bitreverse.*/ctpop.*/ctlz.*/cttz.*`, etc. intrinsics?

**Reply by vnmakarov** [reply]
> @vnmakarov Has libmir an analogous to `llvm.bitreverse.*/ctpop.*/ctlz.*/cttz.*`, etc. intrinsics?

No.  The only non-standard extensions I am working currently to implement are __builtin_expect, __builtin_add_overflow, and __builtin_sub_overflow and some `c2m` specific extensions (attribute antialias, register variables with asm, and some non-standard ABI calls).  That is what I need for good code generation for my  JIT I am implementing with MIR.

I don't think I'll implement intrinsics you mentioned any time soon.  I have too many proposals for MIR which is good as it shows a lot of interests in MIR. I'll implement some proposals but implementing all of them is too much work for me and also I am trying to keep `libmir` small.
